llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -1039,10 +1041,13 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
   // Identify S_ENDPGM instructions which may have to wait for outstanding VMEM
   // stores. In this case it can be useful to send a message to explicitly
   // release all VGPRs before the stores have completed, but it is only safe to
-  // do this if there are no outstanding scratch stores.
+  // do this if:
+  // * there are no outstanding scratch stores
+  // * this is not a callable function
   else if (MI.getOpcode() == AMDGPU::S_ENDPGM ||
            MI.getOpcode() == AMDGPU::S_ENDPGM_SAVED) {
```

Comment:
Do we have to care about callers if we are about to terminate the wave with `S_ENDPGM`? Is this mixing up callees and callers?

Reply:
Ah -- is this about the case where the caller writes to scratch, and the callee terminates the wave, in which case we should not send the message because the caller write might still be pending?

Reply:
Yes! Sorry for being unclear, I'll update the commit message to be more specific.